code/game/objects/items/RSF.dm

```diff
@@ -8,178 +8,200 @@ RSF
+	var/list/allowed_surfaces = list(/turf/open/floor, /obj/structure/table)
+	///The unit of mesure of the matter, for use in text
+	var/discriptor = "fabrication-units"
+	///The verb that describes what we're doing, for use in text
+	var/action_type = "Dispensing"
+
+/obj/item/rsf/Initialize()
+	. = ..()
+	to_despense = cost_by_item[1]
 
 /obj/item/rsf/examine(mob/user)
 	. = ..()
-	. += "<span class='notice'>It currently holds [matter]/30 fabrication-units.</span>"
+	. += "<span class='notice'>It currently holds [matter]/[max_matter] [discriptor].</span>"
 
 /obj/item/rsf/cyborg
 	matter = 30
 
 /obj/item/rsf/attackby(obj/item/W, mob/user, params)
-	if(istype(W, /obj/item/rcd_ammo))
-		if((matter + 10) > 30)
-			to_chat(user, "<span class='warning'>The RSF can't hold any more matter!</span>")
+	if(is_type_in_list(W,matter_by_item))//If the thing we got hit by is in our matter list
+		if(matter > max_matter)
+			to_chat(user, "<span class='warning'>\The [src] can't hold any more [discriptor]!</span>")
 			return
 		qdel(W)
-		matter += 10
+		matter += matter_by_item[W.type]//We add its value
+		if(matter > max_matter)
+			matter = max_matter
 		playsound(src.loc, 'sound/machines/click.ogg', 10, TRUE)
-		to_chat(user, "<span class='notice'>The RSF now holds [matter]/30 fabrication-units.</span>")
-		icon_state = "rsf"
+		to_chat(user, "<span class='notice'>\The [src] now holds [matter]/[max_matter] [discriptor].</span>")
+		icon_state = initial(icon_state)//and set the icon state to the initial value it had
 	else
 		return ..()
 
 /obj/item/rsf/attack_self(mob/user)
+	if(!user)
+		return
 	playsound(src.loc, 'sound/effects/pop.ogg', 50, FALSE)
-	switch(mode)
-		if(5)
-			mode = 1
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Drinking Glass'.</span>")
-		if(1)
-			mode = 2
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Paper'.</span>")
-		if(2)
-			mode = 3
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Pen'.</span>")
-		if(3)
-			mode = 4
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Dice Pack'.</span>")
-		if(4)
-			mode = 5
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Cigarette'.</span>")
+	var/list/item_list
+	var/atom/target = cost_by_item
+	var/cost = 0
+	//Warning, prepare for bodgecode
+	while(istype(target, /list/))//While target is a list we continue the loop
+		item_list = list()
+		for(var/meme in target)//We iterate through all of targets entrys
+			var/atom/test
+			if(istype(meme, /list/))//If it's a list let's use its associated object
+				var/temp = target[meme]
+				test = new temp()
+			else
+				test = new meme()
+			//We then add their data into the radial menu
+			item_list[test.name] = image(icon = test.icon, icon_state = test.icon_state)
+			qdel(test)
+		var/picked = show_radial_menu(user, src, item_list, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE)
```

Comment:
Very interesting way to create `item_list`
but why not make it global with one time populating?
And why not make this way of populating `item_list` is `proc`?

Reply:
I'd prefer to not make it global because it doesn't allow for changes to be made. DItto on the one time populating, that would break sublists as well.

I will make it a proc though, thank you for the advice.